File: Cap3VB/Pilha.vb
Imports System
Imports System.Collections


Public NotInheritable Class Pilha(Of T)
  Implements IEnumerable(Of T)
  Private Const _capacidadeInicial As Integer = 10
  Private _array As T() = New T(_capacidadeInicial - 1) {}
  Private _numeroElementos As Integer = 0

  Public Class PilhaEnumerator
    Implements IEnumerator(Of T)
    Private _array As T()
    Private _index As Integer
    Private _current As T
    Private _parent As Pilha(Of T)

    Friend Sub New(ByVal pilha As Pilha(Of T))
      _parent = pilha
      _array = pilha._array
      _current = Nothing
      _index = -2
    End Sub


    Public Sub Dispose() Implements IEnumerator(Of T).Dispose
      _index = -1
    End Sub

    Private ReadOnly Property CurrentOld() As Object Implements IEnumerator.Current
      Get
        Return Current
      End Get
    End Property


    Private ReadOnly Property Current() As T Implements IEnumerator(Of T).Current
      Get
        If _index < 0 Then
          Throw New InvalidOperationException()
        End If
        Return _current
      End Get
    End Property

    Public Function MoveNext() As Boolean Implements IEnumerator(Of T).MoveNext
      If _index = -2 Then
        ' movenext chamado pela primeira vez
        _index = _parent._numeroElementos
        If _index < 0 Then
          Return False
        End If
        _current = _array(_index)
        Return True
      End If
      If System.Threading.Interlocked.Decrement(_index) <= 0 Then
        Return False
      End If
      _current = _array(_index)
      Return True
    End Function

    Public Sub Reset() Implements IEnumerator(Of T).Reset
      _index = -2
      _current = Nothing
    End Sub
  End Class


  Public Function GetEnumerator() As IEnumerator(Of T) Implements IEnumerable(Of T).GetEnumerator
    Return New PilhaEnumerator(Me)
  End Function

  Private Function GetEnumerator2() As IEnumerator Implements IEnumerable.GetEnumerator
    Return GetEnumerator()
  End Function

  Public ReadOnly Property TotalElementos() As Integer
    Get
      Return _numeroElementos
    End Get
  End Property

  Public Function Pop() As T
    VerificaEstadoPilhaParaLeitura()
    Dim elem As T = _array(System.Threading.Interlocked.Decrement(_numeroElementos))
    _array(_numeroElementos) = Nothing
    Return elem
  End Function

  Private Sub VerificaEstadoPilhaParaLeitura()
    If TotalElementos = 0 Then
      Throw New InvalidOperationException("Não pode executar esta operação sobre uma stack vazia.")
    End If
  End Sub
  Public Sub Push(ByVal elem As T)
    If _array.Length = _numeroElementos Then
      AumentaTamanhoArray()
    End If
    _array(System.Math.Max(System.Threading.Interlocked.Increment(_numeroElementos), _numeroElementos - 1)) = elem
  End Sub

  Private Sub AumentaTamanhoArray()
    Dim novoArray As T() = New T(_array.Length * 2 - 1) {}
    Array.Copy(_array, novoArray, _array.Length)
    _array = novoArray
  End Sub

  Public Function Peek() As T
    VerificaEstadoPilhaParaLeitura()
    Return _array(_numeroElementos - 1)
  End Function
End Class

File: Cap3CSharp/Pilha.cs
using System;
using System.Collections;
using System.Collections.Generic;

namespace Cap3CSharp 
{
    public sealed class Pilha<T> : IEnumerable<T> 
    {
        private const int _capacidadeInicial = 10;
        private T[] _array = new T[_capacidadeInicial];
        private int _numeroElementos = 0;

        public class PilhaEnumerator : IEnumerator<T> 
        {
            private T[] _array;
            private int _index;
            private T _current;
            private Pilha<T> _parent;

            internal PilhaEnumerator(Pilha<T> pilha) 
            {
                _parent = pilha;
                _array = pilha._array;
                _current = default(T);
                _index = -2;
            }

            public T Current 
            {
                get 
                {
                    if (_index < 0) 
                    {
                        throw new InvalidOperationException();
                    }
                    return _current;
                }
            }

            public void Dispose() 
            {
                _index = -1;
            }

            object IEnumerator.Current 
            {
                get { return Current; }
            }

            public bool MoveNext() 
            {
                if (_index == -2) 
                {
                    // movenext chamado pela primeira vez
                    _index = _parent._numeroElementos - 1;
                    if (_index < 0) return false;
                    _current = _array[_index];
                    return true;
                }
                if (--_index < 0) return false;
                _current = _array[_index];
                return true;
            }

            public void Reset() 
            {
                _index = -2;
                _current = default(T);
            }
        }


        public IEnumerator<T> GetEnumerator() 
        {
            return new PilhaEnumerator(this);
        }

        IEnumerator IEnumerable.GetEnumerator() 
        {
            return GetEnumerator();
        }

        public int TotalElementos 
        {
            get { return _numeroElementos; }
        }

        public T Pop() 
        {
            VerificaEstadoPilhaParaLeitura();
            T elem = _array[--_numeroElementos];
            _array[_numeroElementos] = default(T);
            return elem;
        }

        private void VerificaEstadoPilhaParaLeitura() 
        {
            if (TotalElementos == 0) 
            {
                throw new InvalidOperationException(
                  "Não pode executar esta operação sobre uma stack vazia.");
            }
        }
        public void Push(T elem) 
        {
            if (_array.Length == _numeroElementos) 
            {
                AumentaTamanhoArray();
            }
            _array[_numeroElementos++] = elem;
        }

        private void AumentaTamanhoArray() 
        {
            T[] novoArray = new T[_array.Length * 2];
            Array.Copy(_array, novoArray, _array.Length);
            _array = novoArray;
        }

        public T Peek() 
        {
            VerificaEstadoPilhaParaLeitura();
            return _array[_numeroElementos - 1];
        }
    }
}